### PR TITLE
Update spine extension to upstream version from git://github.com/Esoteri...

### PR DESCRIPTION
**Pull request by lumendes**
Sync spine extension with upstream version.
Upstream at: git://github.com/EsotericSoftware/spine-runtimes.git
Synced with revision ab54939e9be79236425be8ad97d406c839bcf953
Only tested in win32 & android platforms.

### CODE REVIEW

**Comment by lumendes** [comment]
Had a look to other pull requests, and apparently there was no need to create a branch for this. Sorry if it's inconvenient.

**Comment by dumganhar** [comment]
Hi, spine parser should be in a namespace(we are now put it in cocos2d::extension).
`c` language has no namespace, so some codes in spine runtime may be conflicted with other modules, for instance, `jsoncpp`.
Therefore, we should change the source code's suffix from `.c` to `.cpp`.

**Comment by lumendes** [comment]
Hi, by spine parser you mean full spine extension or just json files ?
I think all projects should have option to compile C as C++, at least win32 does have.
All .c files contain ifdef for including namespace if compiled as C++.

```
+#ifdef __cplusplus
 namespace cocos2d { namespace extension {
+#endif
```

Does actually renaming .c files to .cpp changes something a compilation level ?

**Comment by dumganhar** [comment]
> I think all projects should have option to compile C as C++, at least win32 does have.

So why not use c++ directly?

**Comment by lumendes** [comment]
I meant all cocos2dx sub-projects, ios, android, etc, in tone of question. I'm not sure if its out-of-the-box, at least in win32 you need to change project settings.
spine upstream dropped the c++ implementation in favor of c only, not sure where to start :)
What is your suggestion ?

**Comment by dumganhar** [comment]
Firstly, I don't know how to use c++ compiler to compile c source in other platforms except win32.
Secondly, namespace is really needed to avoid conflicts.

So IMO, we should change the suffix from `.c` to `.cpp` and change the namespace from `spine` to `cocos2d::extension`.

**Comment by lumendes** [comment]
But that is how it is in this pull request, spine source `.c` and namespace is `cocos2d::extension`
In first comment you said to change `.c` to `.cpp` and now you're saying to change `.cpp` to `.c`.
I'm confused, am I missing something ? :)

**Comment by dumganhar** [comment]
Sorry, my mistake. Update last post.

**Comment by lumendes** [comment]
James, did as you suggested. Compiled fine in win32 and android platforms.
Please have a look into it.

**Comment by dumganhar** [comment]
@lumendes 
Thanks. I'll test it.

File: extensions/spine/Bone.h [hunk 1]
```diff
@@ -28,7 +28,10 @@
 
 #include <spine/BoneData.h>
 
+#ifdef __cplusplus
 namespace cocos2d { namespace extension {
+extern "C" {
+#endif
```
Comment (dumganhar) [inline extensions/spine/Bone.h]: Why all methods in Bone.h should be `extern "C"` while in `Bone.cpp` it doesn't?

**Comment by lumendes** [comment]
@dumganhar yes I believe those were still leftovers from c implementation to allow to compile C as C++.

**Comment by lumendes** [comment]
Latest Travis CI build is passed even if link above still states "Determining merge status"

**Comment by dumganhar** [comment]
Fixed at https://github.com/cocos2d/cocos2d-x/pull/2745.
Thanks. I have added you to AUTHORS.

**Pull request closed without merging**